frb_codegen/src/method_utils.rs

```diff
@@ -0,0 +1,181 @@
+use crate::ir::IrType::{self, StructRef};
+use crate::ir::{IrField, IrTypeStructRef};
+use crate::ir::{IrFile, IrFunc};
+use convert_case::{Case, Casing};
+
+const STATIC_METHOD_MARKER: &str = "__static_method___";
+const METHOD_MARKER: &str = "__method___";
+
+pub(crate) struct MethodNamingUtil;
+
+impl MethodNamingUtil {
+    // Is the function name for a non static method?
+    pub fn is_non_static_method(s: &str) -> bool {
+        s.contains(METHOD_MARKER)
+    }
+
+    // Is the function name for a static method?
+    pub fn is_static_method(s: &str) -> bool {
+        s.contains(STATIC_METHOD_MARKER)
+    }
+
+    // Tests if the function is `f` is a static method for struct with name `struct_name`
+    pub fn is_static_method_for_struct(f: &&IrFunc, struct_name: &str) -> bool {
+        f.name.contains(STATIC_METHOD_MARKER)
+            && f.name.split(STATIC_METHOD_MARKER).last().unwrap() == struct_name
+    }
+
+    //Does `ir_file` has any methods directed for `struct_name`?
+    pub fn has_methods(struct_name: &String, ir_file: &IrFile) -> bool {
+        ir_file.funcs.iter().any(|f| {
+            Self::is_method_for_struct(&f, struct_name)
+                || Self::is_static_method_for_struct(&f, struct_name)
+        })
+    }
+
+    // Tests if the function in `f` is a method for struct with name `struct_name`
+    pub fn is_method_for_struct(f: &&IrFunc, struct_name: &String) -> bool {
+        f.name.contains(METHOD_MARKER) && f.name.split(METHOD_MARKER).last().unwrap() == struct_name
+    }
+
+    // Returns the name of the struct that this method is for
+    pub fn static_method_return_struct_name(s: &str) -> String {
+        s.split(STATIC_METHOD_MARKER).last().unwrap().to_string()
+    }
+
+    // Returns the name of method itself
+    pub fn static_method_return_method_name(s: &str) -> String {
+        s.split(STATIC_METHOD_MARKER).next().unwrap().to_string()
+    }
+
+    // Returns the name of the struct that this method is for
+    pub fn non_static_method_return_struct_name(s: &str) -> String {
+        s.split(METHOD_MARKER).last().unwrap().to_string()
+    }
+
+    // Returns the name of method itself
+    pub fn non_static_method_return_method_name(s: &str) -> String {
+        s.split(METHOD_MARKER).next().unwrap().to_string()
+    }
```

Comment:
same applies here

Reply:
I don't understand. Should I leave them as non pub? Then how to call from the `FunctionName` struct?

Reply:
@lattice0 Maybe FunctionName is within the same file, so non-pub functions can be happily called?

```rust
use std::mem;

struct A;

impl A {
    fn f() {}
}

struct B;

impl B {
    fn g() {
        A::f(); // NOTE it calls non-public
    }
}

fn main() {
}
```

this toy code compiles.